File: code/backsort2.py
#!/usr/bin/env python
#coding: utf-8

import sys

def cat(files):
    for fn in files:
        with open(fn) as f:
            for line in f:
                yield line.rstrip('\n')

def backsort(lines):
    result = {}
    for line in lines:
        result['.'.join(
            reversed(line.split('.')))] = line
    return map(lambda item: item[1],
               sorted(result.items()))

def main():
    if sys.argv[1:]:
        result = backsort(cat(sys.argv[1:]))
    else:
        result = backsort(map(
            lambda line: line.rstrip('\n'),
            sys.stdin))
    for line in result:
        print(line)

if __name__ == '__main__':
    main()

File: code/backsort2.cpp
#include <fstream>          // std::ifstream
#include <functional>       // std::bind/ref
#include <iostream>         // std::cin/cout
#include <istream>          // std::istream
#include <string>           // std::string
#include <vector>           // std::vector
#include <nvwa/file_line_reader.h>
#include <boost/coroutine2/all.hpp>
#include <range/v3/all.hpp>

using namespace std;
using namespace std::placeholders;
using namespace nvwa;
using namespace ranges;

typedef boost::coroutines2::coroutine<const string&> coro_t;

void cat(coro_t::push_type& yield, int files_cnt, char* file_names[])
{
    for (int i = 0; i < files_cnt; ++i) {
        FILE* fp = fopen(file_names[i], "r");
        if (!fp) continue;
        for (auto& line : file_line_reader(fp)) {
            yield(line);
        }
        fclose(fp);
    }
}

template <typename C>
vector<string> backsort(C&& lines)
{
    vector<pair<string, string>> pairs;
    for (const string& line : lines) {
        auto split_line = action::split(line, '.');
        pairs.emplace_back(make_pair(
            to_<string>(split_line | view::reverse | view::join('.')),
            line));
    }
    ranges::sort(pairs);
    return pairs | view::values;
}

int main(int argc, char* argv[])
{
    vector<string> result;
    if (argc > 1) {
        result = backsort(coro_t::pull_type(
            boost::coroutines2::fixedsize_stack(),
            bind(cat, _1, argc - 1, argv + 1)));
    } else {
        result = backsort(file_line_reader(stdin));
    }
    for (auto& item : result) {
       cout << item << endl;
    }
}
